File: android/src/main/java/com/emi/plugins/admob/rewarded/RewardedExecutor.java
package com.emi.plugins.admob.rewarded;

import android.app.Activity;
import androidx.annotation.NonNull;

import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.emi.plugins.admob.AdMobNextGenPlugin;
import com.emi.plugins.admob.ActionCallback;

import com.google.android.libraries.ads.mobile.sdk.common.AdLoadCallback;
import com.google.android.libraries.ads.mobile.sdk.common.AdRequest;
import com.google.android.libraries.ads.mobile.sdk.common.AdValue;
import com.google.android.libraries.ads.mobile.sdk.common.FullScreenContentError;
import com.google.android.libraries.ads.mobile.sdk.common.LoadAdError;
import com.google.android.libraries.ads.mobile.sdk.rewarded.OnUserEarnedRewardListener;
import com.google.android.libraries.ads.mobile.sdk.rewarded.RewardItem;
import com.google.android.libraries.ads.mobile.sdk.rewarded.RewardedAd;
import com.google.android.libraries.ads.mobile.sdk.rewarded.RewardedAdEventCallback;

public class RewardedExecutor {

    private final AdMobNextGenPlugin plugin;
    private RewardedAd mRewardedAd;
    private String currentAdUnitId = "";
    private long lastLoadTime = 0;

    public RewardedExecutor(AdMobNextGenPlugin plugin) {
        this.plugin = plugin;
    }

    public void load(Activity activity, PluginCall call, final ActionCallback callback) {
        String adUnitId = call.getString("adUnitId");
        if (adUnitId == null || adUnitId.isEmpty()) {
            callback.onError("Ad Unit ID is required.");
            return;
        }

        Double retryOpt = call.getDouble("retryInterval");
        long minLoadInterval = (retryOpt != null) ? retryOpt.longValue() : 5000L;
        long currentTime = System.currentTimeMillis();

        if ((currentTime - lastLoadTime) < minLoadInterval) {
            callback.onError("Request too fast. Please wait " + minLoadInterval + "ms to prevent invalid traffic.");
            return;
        }

        this.currentAdUnitId = adUnitId;

        activity.runOnUiThread(() -> {
            RewardedAd.load(
                    new AdRequest.Builder(adUnitId).build(),
                    new AdLoadCallback<RewardedAd>() {
                        @Override
                        public void onAdLoaded(@NonNull RewardedAd rewardedAd) {
                            mRewardedAd = rewardedAd;
                            lastLoadTime = System.currentTimeMillis(); 

                            JSObject ret = new JSObject();
                            ret.put("adUnitId", currentAdUnitId);
                            plugin.notifyListeners("onRewardedAdLoaded", ret);

                            callback.onSuccess();
                        }

                        @Override
                        public void onAdFailedToLoad(@NonNull LoadAdError adError) {
                            mRewardedAd = null;

                            JSObject ret = new JSObject();
                            ret.put("error", adError.getMessage());
                            plugin.notifyListeners("onRewardedAdFailedToLoad", ret);

                            callback.onError(adError.getMessage());
                        }
                    }
            );
        });
    }

    public void show(Activity activity, ActionCallback callback) {
        if (mRewardedAd == null) {
            callback.onError("The rewarded ad is not ready yet.");
            return;
        }

        activity.runOnUiThread(() -> {
            mRewardedAd.setAdEventCallback(new RewardedAdEventCallback() {
                @Override
                public void onAdShowedFullScreenContent() {
                    plugin.notifyListeners("onRewardedAdShowed", new JSObject());
                }

                @Override
                public void onAdDismissedFullScreenContent() {
                    mRewardedAd = null; 
                    plugin.notifyListeners("onRewardedAdDismissed", new JSObject());
                }

                @Override
                public void onAdFailedToShowFullScreenContent(@NonNull FullScreenContentError error) {
                    mRewardedAd = null; 
                    JSObject ret = new JSObject();
                    ret.put("error", error.getMessage());
                    plugin.notifyListeners("onRewardedAdFailedToShow", ret);
                }

                @Override
                public void onAdImpression() {
                    plugin.notifyListeners("onRewardedAdImpression", new JSObject());
                }

                @Override
                public void onAdClicked() {
                    plugin.notifyListeners("onRewardedAdClicked", new JSObject());
                }

                @Override
                public void onAdPaid(@NonNull AdValue value) {
                    JSObject ret = new JSObject();
                    ret.put("adUnitId", currentAdUnitId);
                    ret.put("valueMicros", value.getValueMicros());
                    ret.put("currencyCode", value.getCurrencyCode());
                    ret.put("precisionType", value.getPrecisionType().name());
                    plugin.notifyListeners("onRewardedAdPaid", ret);
                }
            });

            mRewardedAd.show(activity, new OnUserEarnedRewardListener() {
                @Override
                public void onUserEarnedReward(@NonNull RewardItem rewardItem) {
                    JSObject ret = new JSObject();
                    ret.put("amount", rewardItem.getAmount());
                    ret.put("type", rewardItem.getType());
                    plugin.notifyListeners("onRewardedAdReward", ret);
                }
            });
            callback.onSuccess();
        });
    }
}

File: ios/Sources/AdMobNextGenPlugin/rewarded/RewardedExecutor.swift
import Capacitor
import Foundation
import GoogleMobileAds

@objc public class RewardedExecutor: NSObject, FullScreenContentDelegate {

    private weak var plugin: AdMobNextGenPlugin?

    private var rewardedAd: RewardedAd?
    private var currentAdUnitId: String = ""

    private var lastLoadTime: TimeInterval = 0

    private var isRewardEarned: Bool = false

    init(plugin: AdMobNextGenPlugin) {
        self.plugin = plugin
        super.init()
    }

    @objc public func loadAd(_ call: CAPPluginCall) {
        guard let plugin = self.plugin else { return }

        if !plugin.isInitialized {
            call.reject(
                "Google Mobile Ads SDK has not been initialized. Please call initialize() first."
            )
            return
        }

        let adUnitId = call.getString("adUnitId", "")
        if adUnitId.isEmpty {
            call.reject("Ad Unit ID is required.")
            return
        }

        if self.rewardedAd != nil && self.currentAdUnitId == adUnitId {

            var ret = JSObject()
            ret["adUnitId"] = self.currentAdUnitId
            ret["message"] = "Ad already loaded (Cached)"
            plugin.notifyListeners("onRewardedAdLoaded", data: ret)
            call.resolve()
            return
        }

        let retryOpt = call.getDouble("retryInterval")
        let minLoadInterval: TimeInterval =
            (retryOpt != nil) ? (retryOpt! / 1000.0) : 5.0  
        let currentTime = Date().timeIntervalSince1970

        if (currentTime - lastLoadTime) < minLoadInterval {
            call.reject(
                "Request too fast. Please wait \(minLoadInterval * 1000)ms to prevent invalid traffic."
            )
            return
        }

        self.currentAdUnitId = adUnitId

        let request = Request()

        Task { @MainActor in
            do {

                let ad = try await RewardedAd.load(
                    with: adUnitId,
                    request: request
                )

                self.rewardedAd = ad
                self.rewardedAd?.fullScreenContentDelegate = self

                self.rewardedAd?.paidEventHandler = { [weak self] value in
                    guard let self = self else { return }
                    var ret = JSObject()
                    ret["adUnitId"] = self.currentAdUnitId
                    ret["valueMicros"] = value.value.stringValue
                    ret["currencyCode"] = value.currencyCode
                    ret["precisionType"] = value.precision.rawValue
                    self.plugin?.notifyListeners("onRewardedAdPaid", data: ret)
                }

                self.lastLoadTime = Date().timeIntervalSince1970  

                var ret = JSObject()
                ret["adUnitId"] = self.currentAdUnitId
                plugin.notifyListeners("onRewardedAdLoaded", data: ret)

                call.resolve()

            } catch {

                var ret = JSObject()
                ret["error"] = error.localizedDescription
                plugin.notifyListeners("onRewardedAdFailedToLoad", data: ret)

                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func showAd(_ call: CAPPluginCall) {
        guard let plugin = self.plugin else { return }

        if self.rewardedAd == nil {
            call.reject("The rewarded ad is not ready yet.")
            return
        }

        DispatchQueue.main.async {
            guard let viewController = plugin.bridge?.viewController else {
                call.reject("Root ViewController not found.")
                return
            }

            if let ad = self.rewardedAd {
                self.isRewardEarned = false

                ad.present(from: viewController) { [weak self] in
                    self?.isRewardEarned = true
                    let reward = ad.adReward
                    var ret = JSObject()
                    ret["amount"] = reward.amount.doubleValue  
                    ret["type"] = reward.type
                    self?.plugin?.notifyListeners(
                        "onRewardedAdReward",
                        data: ret
                    )
                }
                call.resolve()
            }
        }
    }

    public func adWillPresentFullScreenContent(_ ad: FullScreenPresentingAd) {
        plugin?.notifyListeners("onRewardedAdShowed", data: [:])
    }

    public func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        rewardedAd = nil  

        if !self.isRewardEarned {
            plugin?.notifyListeners("onRewardedAdSkip", data: [:])
        }
        plugin?.notifyListeners("onRewardedAdDismissed", data: [:])
    }

    public func ad(
        _ ad: FullScreenPresentingAd,
        didFailToPresentFullScreenContentWithError error: Error
    ) {
        rewardedAd = nil  
        var ret = JSObject()
        ret["error"] = error.localizedDescription
        plugin?.notifyListeners("onRewardedAdFailedToShow", data: ret)
    }

    public func adDidRecordImpression(_ ad: FullScreenPresentingAd) {
        plugin?.notifyListeners("onRewardedAdImpression", data: [:])
    }

    public func adDidRecordClick(_ ad: FullScreenPresentingAd) {
        plugin?.notifyListeners("onRewardedAdClicked", data: [:])
    }
}
